Fixes in documentation

ligurio opened the pull request:
- Fixed a typo
- Add note about functions that returns more than one iterator (#57)
- Removed param x in description of `partition()`
- Replaced http with https links where it is possible

Totktonada commented:
I clarified how it should be fixed in the issue: https://github.com/luafun/luafun/issues/57#issuecomment-948200244 (because actual discussion was already there).

ligurio commented:
> I clarified how it should be fixed in the issue: #57 (comment) (because actual discussion was already there).

Thanks for detailed comment. I have updated commit related to fix #57. However I did it in another way: added note about multireturn to "Getting started" chapter and mention not in descriptions for `partition` and `span`.

ligurio commented:
New changes rendered fine in documentation:

![Screenshot from 2021-11-24 15-43-46](https://user-images.githubusercontent.com/1151557/143240832-50bef40a-f059-41ae-b136-48aef51e4121.png)

Totktonada commented on doc/under_the_hood.rst:
```diff
@@ -145,10 +145,10 @@ tail-, up- and down-recursion [#luajit-recursion]_, unroll compositions of
+.. [#pure_function] https://en.wikipedia.org/wiki/Pure_function
+.. [#turing] `Proved by Turing <https://en.wikipedia.org/wiki/Halting_problem>`_
+.. [#tracing_jit] https://en.wikipedia.org/wiki/Tracing_just-in-time_compilation
+.. [#luajit-recursion] https://lambda-the-ultimate.org/node/3851#comment-57679
```
It has a self-signed certificate, it is better to leave `http://` here.

ligurio replied:
Fixed.

Totktonada commented on doc/filtering.rst:
```diff
@@ -96,7 +96,7 @@ This section contains functions to filter values during iteration.
    :returns: {gen1, param1, state1}, {gen2, param2, state2}
 
    The function returns two iterators where elements do and do not satisfy the
```
A general comment: I would prefix all commits with `doc:` prefix, so it'll nicely look in git log.

ligurio replied:
I have reviewed previous commits and all of them doesn't have prefixes, so I decided to write a similar commit header.
Ok, let's use prefixes.

Totktonada commented on doc/filtering.rst:
```diff
@@ -95,7 +95,7 @@ This section contains functions to filter values during iteration.
    :returns: {gen1, param1, state1}, {gen2, param2, state2}
```
It is still incorrect. Please look at the 'Actual mistakes' section in my comment.

ligurio replied:
```
--- a/doc/filtering.rst
+++ b/doc/filtering.rst
@@ -92,10 +92,10 @@ This section contains functions to filter values during iteration.
 .. function:: partition(predicate, gen, param, state)
               iterator:partition(predicate)
 
-   :returns: {gen1, param1, state1}, {gen2, param2, state2}
+   :returns: iterator, iterator
 
```

Totktonada commented on doc/getting_started.rst:
```diff
@@ -235,7 +235,41 @@ Now you can use **Lua Fun**:
     one
 
-Please note that functions support multireturn.
+Please note that functions support multireturn. It means that some of them may
+return more than one iterator, for example functions ``partition`` and ``span``.
```
I guess the idea of the 'functions support multireturn' sentence was about all functions: they return `gen, param, state`. A note about functions that returns two iterators is a separate thing.

ligurio replied:
Fixed.

```asciidoc
--- a/doc/getting_started.rst
+++ b/doc/getting_started.rst
@@ -235,11 +235,13 @@ Now you can use **Lua Fun**:
     2 c
     one
 
-Please note that functions support multireturn. It means that some of them may
-return more than one iterator, for example functions ``partition`` and ``span``.
+Please note that functions support multireturn - many functions return a
+triplet ``gen``, ``param`` and ``state``, so called :ref:`iterator triplet
+<iterator_triplet>`.
 
-One call ``:unwrap()`` on generators to pass them into the ``for <..> in <...>``
-loop. Example:
+Please note that functions support returns of two iterators. Example of these
+functions are :func:`partition` and :func:`span`. One can call ``:unwrap()``
+on iterators to pass them into the ``for <..> in <...>`` loop. Example:
 
 .. code-block:: bash
 
```

Totktonada commented on doc/getting_started.rst:
```diff
@@ -235,7 +235,41 @@ Now you can use **Lua Fun**:
+Please note that functions support multireturn. It means that some of them may
+return more than one iterator, for example functions ``partition`` and ``span``.
+
+One call ``:unwrap()`` on generators to pass them into the ``for <..> in <...>``
+loop. Example:
```
I would prefer to call luafun iterators as 'iterators', not generators. At least to don't confuse with a subset of iterators that are generators. And to don't confuse with just `gen` function.

BTW, 'one call :unwrap()' seems lacks a verb if I don't miss anything.

ligurio replied:
Fixed.

```asciidoc
--- a/doc/getting_started.rst
+++ b/doc/getting_started.rst
@@ -238,8 +238,8 @@ Now you can use **Lua Fun**:
 Please note that functions support multireturn. It means that some of them may
 return more than one iterator, for example functions ``partition`` and ``span``.
 
-One call ``:unwrap()`` on generators to pass them into the ``for <..> in <...>``
-loop. Example:
+One can call ``:unwrap()`` on iterators to pass them into the ``for <..> in
+<...>`` loop. Example:
 
 .. code-block:: bash
 
```

ligurio replied:
![Screenshot from 2021-11-29 21-21-11](https://user-images.githubusercontent.com/1151557/143921990-d16825fc-136d-4412-9bb3-7c5bdaa02d2a.png)

Totktonada commented on doc/getting_started.rst:
```diff
@@ -235,7 +235,41 @@ Now you can use **Lua Fun**:
+
+.. code-block:: bash
+
+   it1, it2 = fun.range(10):partition(function(i) return i % 2 == 0 end)
+   for i in it1:unwrap() do print(i) end
```
Can we add angle brackets as prompt to mark input lines? As I see, it is used above.

ligurio replied:
Sure, added.

Totktonada commented on doc/getting_started.rst:
```diff
@@ -235,7 +235,41 @@ Now you can use **Lua Fun**:
     one
 
-Please note that functions support multireturn.
+Please note that functions support multireturn. It means that some of them may
```
Just thought: is not it too hard to be in the 'Getting started' section?

ligurio replied:
I think yes. There is another appropriate section - 'Under the hood', but it looks as additional information for advanced users. Note about multireturn is for everyone who use `span` or `partititon`.

Totktonada commented on doc/filtering.rst:
```diff
@@ -95,7 +95,7 @@ This section contains functions to filter values during iteration.
 
    The function returns two iterators where elements do and do not satisfy the
-   predicate. Equivalent to:
+   predicate. See note about multireturn in :doc:`getting_started`. Equivalent to:
```
I hope there is a way to set an anhor and link it here. Looks quite unconvenient to find needed recipe over the doc, when you don't know what exactly was meant.

(I know, it was deliberate decision, but) I would place the right here (somewhere below examples) and link from `span`. I don't mind a general note and linking it from both functions, but I see no appropriate place for such note.

ligurio replied:
> I hope there is a way to set an anhor and link it here. Looks quite unconvenient to find needed recipe over the doc, when you don't know what exactly was meant.

Added anchor.

Totktonada commented on doc/getting_started.rst:
```diff
@@ -19,11 +19,11 @@ e.g. Tarantool_ or OpenResty_.
+.. _Tarantool: https://tarantool.org/
+.. _OpenResty: https://openresty.org/
+.. _diff: https://en.wikipedia.org/wiki/Diff
+.. _sphinx: https://sphinx-doc.org/
```
This link does not work, but https://www.sphinx-doc.org works.

Totktonada commented on doc/filtering.rst:
```diff
@@ -92,11 +92,11 @@ This section contains functions to filter values during iteration.
+   predicate. See note about multireturn in :ref:`Getting started
+   <multi_return>`. Equivalent to:
 
    .. code-block:: lua
```
See below:

> Iterators especially returned in tables to work with :func:`zip` and other functions.

And see 'Actual mistakes' section in my comment in the issue. In short: now all iterators are triplets with a table in first item, this comment is just irrelevant now.

Totktonada commented on doc/filtering.rst:
```diff
@@ -92,11 +92,11 @@ This section contains functions to filter values during iteration.
 
-   :param x: a value to find
-   :returns: {gen1, param1, state1}, {gen2, param2, state2}
+   :returns: iterator, iterator
```
We agreed that `iterator` is correct term for a return value and `{gen, param, state}` is incorrect one (see 'Correctness' section of my comment in the issue). But description of the `cycle()` function was not updated and still tell us about `{gen, param, state}`.

Totktonada commented on doc/getting_started.rst:
```diff
@@ -235,7 +235,45 @@ Now you can use **Lua Fun**:
-Please note that functions support multireturn.
+.. _multi_return:
+
+Please note that functions support multireturn - many functions return a
+triplet ``gen``, ``param`` and ``state``, so called :ref:`iterator triplet
+<iterator_triplet>`.
+
+Please note that functions support returns of two iterators. Example of these
+functions are :func:`partition` and :func:`span`. One can call ``:unwrap()``
+on iterators to pass them into the ``for <..> in <...>`` loop. Example:
```
Will not 'one can call `:unwrap()` on iterators to pass then into `for <..> in <..>` loop' look as a general note instead of be a recipe specifically for functions that return two iterators?

To be honest, I would move accents here and would write something like the following.

<details>
<summary>Example of wording</summary>

> Please note that luafun functions employ multireturn feature of Lua language and return `gen`, `param`, `state` :ref:\`triplet <iterator_triplet>\`. This is exactly what a \`\`for\`\` loop expects, so luafun functions can be used in the place where \`\`pairs\`\` or \`\`ipairs\`\` is usually used:
>
> ```lua
> -- Usual usage.
> > for i in range(3) do print(i) end
> 1
> 2
> 3
> -- The same, but in the more explicit form.
> > gen, param, state = range(3)
> > for i in gen, param, state do print(i) end
> 1
> 2
> 3
> ```
>
> The main practical idea here is that the following will be incorrect:
>
> ```lua
> > it = range(3)
> > for i in it do print(i) end -- error
> ```
>
> However \`\`param\`\` and \`\`state\`\` are actually wrapped inside the \`\`it\`\` iterator and it is possible to unwrap them back to \`\`gen\`\`, \`\`param\`\`, \`\`state\`\`:
>
> ```lua
> -- Explicit form.
> > it = range(3)
> > gen, param, state = it:unwrap()
> > for i in gen, param, state do print(i) end
> 1
> 2
> 3
> -- Usual simple form.
> > it = range(3)
> > for i in it:unwrap() do print(i) end
> 1
> 2
> 3
> ```
>
> Usually it is not important: just call `:totable()` or pass the returned triplet into a \`\`for\`\` loop. But the \`\`unwrap()\`\` trick is very useful for luafun functions, which return two iterators (such as :func:\`partition\` and :func:\`span\`). Consider the example:
>
> \<...\>
</details>

BTW, just observed that the `unwrap()` function/method is not documented.

Do you see my idea? 'luafun supports multireturn' means nothing, the whole Lua language supports it. The practical sense is that we can use the return values in a `for` loop directly. And that we should save them into three variables to pass somewhere or should call `unwrap()`. Next, 'one can call `unwrap()`' means nothing: the practical sense is that one *should* use it to pass iterators from `partition` and `span` into a `for` loop, at least on the first iterator.

I hope it does not sound as 'I like this wording and don't like that wording'. If my idea is not clear, let's discuss it voicely.
